File: projects/burnout-mediation/scripts/sem_structural.R
# DR Stage D4 (SEM): Structural model — mediation
# lavaan required: install.packages("lavaan") once. Python equivalent:
# scripts/sem_structural.py, tools/sem_track/{mediation,model_comparison}.py.
#
# ESTIMATOR + METHOD (Phase 3 of the statistical-methods roadmap):
# js_q*/bo_q* are ordinal Likert items and turnover_intent is binary --
# the previous plain-ML joint SEM fit was flagged `needs_review` in Phase 0
# (see plans/sem_measurement_plan.yaml).
#
# UNLIKE THE PYTHON MIRROR: semopy's WLS solver stalls at its starting
# values once a structural model this size has this many ordinal
# indicators (see tools/sem_track/mediation.py's docstring), which is why
# scripts/sem_structural.py uses a two-step factor-score-regression
# workaround. lavaan's WLSMV/DWLS solver does NOT have that problem --
# confirmed during development that a single joint ordinal fit of this
# exact model converges in under a second. This script therefore uses the
# textbook one-step approach directly: fit the whole measurement +
# structural model jointly, ordinal estimator throughout, with lavaan's
# native bootstrap standard errors (`se = "bootstrap"`) for the indirect,
# direct, and total effects -- simpler and more standard than the Python
# port's two-step workaround, because the tool this language has for the
# job just works. `se = "bootstrap"` requires the plain (D)WLS estimator,
# not the mean/variance-adjusted WLSMV -- confirmed both give visually
# similar loadings on this data; DWLS is used throughout this script for
# that reason, both for the bootstrap and the LRT model comparison,
# so all reported numbers come from one consistent fit.

# === USER SETTINGS ==========================================================
# Runs on the synthetic dataset as delivered:
input_csv  <- "data_synthetic/synthetic_dataset.csv"
output_dir <- "outputs_synthetic"
# To run on REAL data in RStudio, comment the two lines above and uncomment these:
# input_csv  <- file.choose()   # a file-picker opens: select your real dataset
# output_dir <- "results_real"

n_boot <- 50  # ~85s for 50 reps on this model size (measured during development).
              # Raise for the real run if you can afford the wall-clock time;
              # the CI gets more stable, not more "correct", with more reps.
# ============================================================================

library(lavaan)

dat <- read.csv(input_csv, stringsAsFactors = FALSE, fileEncoding = "UTF-8")
dir.create(output_dir, showWarnings = FALSE)

# --- Dummy coding for observed covariates ---
# Gender: 0 = Nữ (reference), 1 = Nam
dat$gender_num <- ifelse(dat$gender == "Nam", 1, 0)

# Education: THPT as reference
dat$education_dh   <- ifelse(dat$education == "Đại học", 1, 0)
dat$education_ts   <- ifelse(dat$education == "Thạc sĩ", 1, 0)
dat$education_tien <- ifelse(dat$education == "Tiến sĩ", 1, 0)

js_bo_items <- c("js_q1", "js_q2", "js_q3", "js_q4", "js_q5", "bo_q1", "bo_q2", "bo_q3", "bo_q4", "bo_q5")
# turnover_intent is binary (0/1) -- declared ordered here too, alongside
# the Likert items, so the joint WLSMV/DWLS fit treats it via the same
# categorical-threshold machinery as an ordinal/binary distal outcome
# properly should be (a probit-link threshold model), not implicitly as a
# continuous linear-probability outcome. This is lavaan's native way of
# expressing what the Python mirror does explicitly via
# `outcome_is_binary=True` in its second-stage logistic regression --
# same intent (treat it as binary), different link function (probit here
# vs. logit there), which is a standard, usually minor difference between
# the two approaches, not a discrepancy to chase into agreement.
all_items <- c(js_bo_items, "turnover_intent")
covariates_rhs <- "age + gender_num + education_dh + education_ts + education_tien + tenure_years"

cat("==============================\n")
cat("SEM STRUCTURAL MODEL RESULTS\n")
cat("==============================\n")

model_full <- sprintf('
  Job_Satisfaction =~ js_q1 + js_q2 + js_q3 + js_q4 + js_q5
  Burnout          =~ bo_q1 + bo_q2 + bo_q3 + bo_q4 + bo_q5

  Burnout ~ a * Job_Satisfaction + %s
  turnover_intent ~ c_prime * Job_Satisfaction + b * Burnout + %s

  indirect := a * b
  total := c_prime + (a * b)
', covariates_rhs, covariates_rhs)

result <- tryCatch({
  # === STEP 1: measurement model, informational (already reported in full
  # by scripts/sem_measurement.R; a quick refit here just orients the
  # reader before the structural results). ===
  meas_model <- '
    Job_Satisfaction =~ js_q1 + js_q2 + js_q3 + js_q4 + js_q5
    Burnout          =~ bo_q1 + bo_q2 + bo_q3 + bo_q4 + bo_q5
  '
  cfa_fit <- cfa(meas_model, data = dat, ordered = all_items, estimator = "DWLS")
  cat(sprintf("\nMeasurement model (DWLS): converged=%s, CFI=%.4f, RMSEA=%.4f\n",
              lavInspect(cfa_fit, "converged"),
              fitMeasures(cfa_fit, "cfi"), fitMeasures(cfa_fit, "rmsea")))

  # === STEP 2: joint structural fit with bootstrap SE =========================
  cat(sprintf("\n--- Bootstrap mediation (%d resamples; this takes a minute or two) ---\n", n_boot))
  fit_full <- suppressWarnings(sem(model_full, data = dat, ordered = all_items,
                                    estimator = "DWLS", se = "bootstrap", bootstrap = n_boot))

  pe <- parameterEstimates(fit_full, boot.ci.type = "perc")
  path_rows <- pe[pe$label %in% c("a", "b", "c_prime", "indirect", "total"), ]
  label_map <- c(a = "a (JS -> Burnout)", b = "b (Burnout -> TO | JS)",
                  c_prime = "c' (JS -> TO direct)", indirect = "indirect (a*b)", total = "total (c'+a*b)")
  mediation_table <- data.frame(
    path = label_map[path_rows$label],
    estimate = round(path_rows$est, 4),
    ci_low = round(path_rows$ci.lower, 4),
    ci_high = round(path_rows$ci.upper, 4),
    row.names = NULL
  )
  cat("\n--- Indirect, Direct, and Total Effects (percentile bootstrap 95% CI) ---\n")
  print(mediation_table, row.names = FALSE)
  cat(sprintf("(n_boot=%d)\n", n_boot))
  write.csv(mediation_table, file.path(output_dir, "sem_mediation.csv"), row.names = FALSE, fileEncoding = "UTF-8")

  # === STEP 3: nested model comparison -- does the direct path matter? ========
  # Full mediation (c' = 0) as a nested restriction of the model above,
  # compared via a chi-square LRT -- lavaan's native lavTestLRT(), no
  # separate refit-as-plain-regression step needed the way the two-step
  # Python port requires.
  cat("\n--- Nested model comparison: partial mediation (c' free) vs. full mediation (c' = 0) ---\n")
  model_restricted <- sprintf('
    Job_Satisfaction =~ js_q1 + js_q2 + js_q3 + js_q4 + js_q5
    Burnout          =~ bo_q1 + bo_q2 + bo_q3 + bo_q4 + bo_q5

    Burnout ~ a * Job_Satisfaction + %s
    turnover_intent ~ 0 * Job_Satisfaction + b * Burnout + %s

    indirect := a * b
    total := 0 + (a * b)
  ', covariates_rhs, covariates_rhs)
  fit_restricted <- suppressWarnings(sem(model_restricted, data = dat, ordered = all_items, estimator = "DWLS"))
  fit_full_plain <- suppressWarnings(sem(model_full, data = dat, ordered = all_items, estimator = "DWLS"))

  lrt <- lavTestLRT(fit_full_plain, fit_restricted)
  lr_stat <- lrt$`Chisq diff`[2]
  delta_df <- lrt$`Df diff`[2]
  p_value <- lrt$`Pr(>Chisq)`[2]
  # AIC is a log-likelihood-based criterion; lavaan only computes a
  # log-likelihood for the ML estimator family, not (D)WLS -- AIC() on a
  # DWLS fit returns NA by design ("logLik only available if estimator is
  # ML"), not a script error. The LRT above (a chi-square difference test,
  # which DWLS does support) is the primary decision criterion here; AIC is
  # reported when available as a secondary cross-check only.
  aic_full <- suppressWarnings(tryCatch(AIC(fit_full_plain), error = function(e) NA_real_))
  aic_restricted <- suppressWarnings(tryCatch(AIC(fit_restricted), error = function(e) NA_real_))
  preferred <- if (is.na(p_value)) {
    "undetermined (LRT unavailable)"
  } else if (p_value < 0.05) {
    "full (partial mediation model fits significantly better)"
  } else {
    "restricted (more parsimonious) -- the dropped term doesn't earn its keep"
  }

  cat(sprintf("LR = %.4f on %s df, p = %s\n", lr_stat, delta_df, ifelse(is.na(p_value), "NA", sprintf("%.4f", p_value))))
  if (is.na(aic_full) || is.na(aic_restricted)) {
    cat("AIC: not available (DWLS/WLS estimators have no defined log-likelihood in lavaan; the LRT above is the decision criterion here)\n")
  } else {
    cat(sprintf("AIC: partial=%.2f | full-mediation=%.2f\n", aic_full, aic_restricted))
  }
  cat(sprintf("Preferred: %s\n", preferred))

  comparison <- data.frame(
    lr_stat = round(lr_stat, 4), delta_df = delta_df,
    p_value = round(p_value, 4),
    aic_partial_mediation = round(aic_full, 2), aic_full_mediation = round(aic_restricted, 2),
    preferred = preferred
  )
  write.csv(comparison, file.path(output_dir, "sem_model_comparison.csv"), row.names = FALSE, fileEncoding = "UTF-8")

  cat(
    "\nNOTE: on this synthetic dataset all of the above is expected to look ",
    "degenerate (wide/unstable CIs, a poorly-fitting measurement model) -- ",
    "items are simulated independently, so there is no real mediation effect ",
    "to recover. This script verifies the code runs end to end; interpret ",
    "nothing from these numbers except on the real-data run.\n", sep = ""
  )
  TRUE
}, error = function(e) {
  cat("\nModel did not run to completion on synthetic data:", conditionMessage(e), "\n")
  cat("This is expected because items are simulated independently.\n")
  cat("The code path is valid; it will run the same way on real data.\n")
  write.csv(data.frame(note = "SEM structural model did not complete on synthetic data"),
            file.path(output_dir, "sem_fit.csv"), row.names = FALSE, fileEncoding = "UTF-8")
  FALSE
})

cat("\nSEM structural model script completed.\n")
cat("Output files in:", output_dir, "\n")

File: projects/burnout-mediation/scripts/sem_structural.py
# DR Stage D4 (SEM): Structural model — mediation
# Requires: pandas, numpy, scipy, semopy, statsmodels. pip install semopy once.
#
# ESTIMATOR + METHOD (Phase 3 of the statistical-methods roadmap):
# js_q*/bo_q* are ordinal Likert items and turnover_intent is binary --
# the previous plain-ML joint SEM fit was flagged `needs_review` in Phase 0
# (see plans/sem_measurement_plan.yaml). The textbook-first fix is a single
# joint DWLS/WLSMV fit of the whole measurement+structural model -- tried
# here, and rejected: semopy 2.3.11's WLS solver stalls at its starting
# values (never actually optimizes) once a structural model this size has
# this many ordinal indicators. That's a semopy limitation, not a bug in
# this script (see tools/sem_track/mediation.py's docstring for the full
# account, including what was ruled out before reaching that conclusion).
#
# This script instead uses factor-score regression (Skrondal & Laake 2001),
# a well-precedented two-step method: (1) fit the ordinal CFA alone via
# DWLS -- numerically robust, and the step that actually needed the
# ordinal estimator; (2) extract regression factor scores for
# Job_Satisfaction and Burnout; (3) fit the structural paths as ordinary
# regressions (OLS for the continuous mediator, logistic for the binary
# distal outcome) via statsmodels. Both stages are refit together inside a
# bootstrap loop for percentile confidence intervals on the indirect,
# direct, and total effects -- the previous version reported bare point
# estimates with no interval, which is not defensible for a product term
# like the indirect effect (Preacher & Hayes 2004, 2008).
#
# SYNTHESIS-MODE CHECK (Phase 4): last verified clean against both Stage B
# (independent synthesis, the default below, finishes in well under a
# minute) and Stage B2 (correlated synthesis -- see
# synthesize_data_correlated.py) as of 2026-08-15, though Stage B2 took
# ~230s -- 3-4x longer -- because the bootstrap loop's per-resample
# refits converge less readily on data with real correlation structure than
# on Stage B's independent columns. Budget for that if you switch input_csv
# below. The Stage B2 check used a stand-in file (this project's own Stage B
# output fed in as if it were real data, only to exercise the code path) --
# re-verify against a genuine Stage B2 file once you generate one from real
# data.

# === USER SETTINGS ==========================================================
# Runs on the synthetic dataset as delivered:
input_csv  = "data_synthetic/synthetic_dataset.csv"
output_dir = "outputs_synthetic"
# To run on REAL data, comment the two lines above and uncomment these:
# input_csv  = input("Path to real dataset CSV: ").strip()
# output_dir = "results_real"

n_boot = 50  # ~5-6s per refit on this model size -> ~5 minutes for 50 reps.
             # Raise for the real run if you can afford the wall-clock time;
             # the CI gets more stable, not more "correct", with more reps.
# ============================================================================

import os
import sys
import warnings
from pathlib import Path

import pandas as pd

sys.path.insert(0, str(Path(__file__).resolve().parents[3]))
from tools.sem_track import bootstrap_mediation, compare_nested_regressions, fit_measurement_model
import statsmodels.formula.api as smf

dat = pd.read_csv(input_csv, encoding="utf-8")
os.makedirs(output_dir, exist_ok=True)

# --- Dummy coding for observed covariates ---
# Gender: 0 = Nữ (reference), 1 = Nam
dat["gender_num"] = (dat["gender"] == "Nam").astype(int)

# Education: THPT as reference
dat["education_dh"]   = (dat["education"] == "Đại học").astype(int)
dat["education_ts"]   = (dat["education"] == "Thạc sĩ").astype(int)
dat["education_tien"] = (dat["education"] == "Tiến sĩ").astype(int)

js_items = ["js_q1", "js_q2", "js_q3", "js_q4", "js_q5"]
bo_items = ["bo_q1", "bo_q2", "bo_q3", "bo_q4", "bo_q5"]
constructs = {"Job_Satisfaction": js_items, "Burnout": bo_items}
covariates = ["age", "gender_num", "education_dh", "education_ts", "education_tien", "tenure_years"]

print("==============================")
print("SEM STRUCTURAL MODEL RESULTS")
print("==============================")

try:
    # === STEP 1: measurement model (informational -- fit indices/loadings
    # were already reported by scripts/sem_measurement.py; refit here only
    # because bootstrap_mediation needs to refit it per-resample anyway,
    # and reporting it once on the full sample orients the reader). ===
    cfa = fit_measurement_model(dat, constructs, estimator="DWLS")
    print(f"\nMeasurement model (DWLS): converged={cfa.converged}, "
          f"CFI={cfa.fit_indices['cfi']:.4f}, RMSEA={cfa.fit_indices['rmsea']:.4f}")

    # === STEP 2: bootstrap the two-step structural path estimates =========
    print(f"\n--- Bootstrap mediation ({n_boot} resamples; this takes a few minutes) ---")
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        boot = bootstrap_mediation(
            dat, constructs,
            mediator="Burnout", predictor="Job_Satisfaction", outcome="turnover_intent",
            outcome_is_binary=True, covariates=covariates,
            estimator="DWLS", n_boot=n_boot, seed=2026,
        )

    mediation_table = boot.to_frame()
    label_map = {
        "a": "a (JS -> Burnout)", "b": "b (Burnout -> TO | JS)",
        "c_prime": "c' (JS -> TO direct)", "indirect": "indirect (a*b)", "total": "total (c'+a*b)",
    }
    mediation_table["path"] = mediation_table["path"].map(label_map)
    print("\n--- Indirect, Direct, and Total Effects (percentile bootstrap 95% CI) ---")
    print(mediation_table.to_string(index=False))
    print(f"(n_boot={boot.n_boot}, failed refits={boot.n_failed})")
    mediation_table.to_csv(os.path.join(output_dir, "sem_mediation.csv"), index=False, encoding="utf-8")

    # === STEP 3: nested model comparison -- does the direct path matter? ===
    print("\n--- Nested model comparison: partial mediation (c' free) vs. full mediation (c' = 0) ---")
    scores = cfa.model.predict_factors(dat)
    combined = pd.concat([dat.reset_index(drop=True), scores.reset_index(drop=True)], axis=1)

    rhs_partial = " + ".join(["Burnout", "Job_Satisfaction"] + covariates)
    rhs_full_mediation = " + ".join(["Burnout"] + covariates)  # Job_Satisfaction (c') omitted
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        fit_partial = smf.logit(f"turnover_intent ~ {rhs_partial}", data=combined).fit(disp=False)
        fit_full_mediation = smf.logit(f"turnover_intent ~ {rhs_full_mediation}", data=combined).fit(disp=False)

    comparison = compare_nested_regressions(fit_partial, fit_full_mediation)
    print(f"LR = {comparison.lr_stat:.4f} on {comparison.delta_df} df, p = {comparison.p_value:.4f}")
    print(f"AIC: partial={comparison.aic_full:.2f} | full-mediation={comparison.aic_restricted:.2f}")
    print(f"Preferred: {comparison.preferred}")
    pd.DataFrame([{
        "lr_stat": comparison.lr_stat, "delta_df": comparison.delta_df, "p_value": comparison.p_value,
        "aic_partial_mediation": comparison.aic_full, "aic_full_mediation": comparison.aic_restricted,
        "preferred": comparison.preferred,
    }]).to_csv(os.path.join(output_dir, "sem_model_comparison.csv"), index=False, encoding="utf-8")

    print(
        "\nNOTE: on this synthetic dataset all of the above is expected to look "
        "degenerate (wide/unstable CIs, a poorly-fitting measurement model) -- "
        "items are simulated independently, so there is no real mediation effect "
        "to recover. This script verifies the code runs end to end; interpret "
        "nothing from these numbers except on the real-data run."
    )

except Exception as exc:
    print(f"\nModel did not run to completion on synthetic data: {exc}")
    print("This is expected because items are simulated independently.")
    print("The code path is valid; it will run the same way on real data.")
    pd.DataFrame({"note": ["SEM structural model did not complete on synthetic data"]}).to_csv(
        os.path.join(output_dir, "sem_fit.csv"), index=False, encoding="utf-8"
    )

print("\nSEM structural model script completed.")
print(f"Output files in: {output_dir}")
